Documentation updates for v7.0.0 (#348)
auth0/auth0.net: docs-source/migrating-to-v4.md

# Migration Guide

## Migrating from v3 to v4

> Version 4 of both the Authentication API SDK as well as the Management API SDK are major breaking changes. This document will discuss the reason for these changes, as well as instructions on how to migrate to the new version.

Reviewer: ```suggestion
Version 4 of both the Authentication API SDK as well as the Management API SDK include breaking changes. This document will discuss the reason for these changes, as well as instructions on how to migrate to the new version.
```

### OIDC Conformance

> The main reason for the breaking changes is related to the OIDC-conformance which was added to the Authentication API. Because of this, the behaviour of some of the existing Authentication API endpoints have changed, and other endpoints are being deprecated.

Reviewer: ```suggestion
The main reason for the breaking changes is related to improved OIDC compliance which was added to the Authentication API. Because of this, the behaviour of some of the existing Authentication API endpoints have changed, and other endpoints are being deprecated.
```

> For a full background and other details please refer to the official [OIDC-conformant authentication migration guide](https://auth0-docs-content-pr-3100.herokuapp.com/docs/api-auth/tutorials/migration).

Reviewer: ```suggestion
For a full background and other details please refer to the official [OIDC-conformant authentication migration guide](https://auth0.com/docs/api-auth/tutorials/adoption).
```

Where did that Heroku link come from? I'm wondering if that's part of another document out there.

Author: The v4 and v6 migration guides were already written I just split them into their own pages so not sure where it came from.

### Better separation of the Authentication and Management API

Because of the changes to the Authentication pipeline for OIDC conformance, some breaking changes were introduced and we are therefore required to increase the version number of the Authentication API SDK. One problem however was that some classes were shared in between the Authentication API SDK and the Management API SDK, in particular the information returned from the `/userinfo` endpoint. 

In the new OIDC conformant pipeline this is not the case anymore, as the endpoint return claims which conform to the [OIDC standard](http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims).

Because of this single instance of shared data between the Authentication and Management API SDKs, a lot of the _Models_ returned by the various methods has been stored in the [Auth0.Core NuGet package](https://www.nuget.org/packages/Auth0.Core). With version 4 this is not the case anymore. 

> All model classes are now stored in the Nuget package which they relate to. So all Authentication API SDK model classes are stored in the [Auth0.AuthenticationAPI NuGet package](https://www.nuget.org/packages/Auth0.AuthenticationApi). Likewise, all Management API SDK model classes are stored in the [Auth0.ManagementAPI NuGet package](https://www.nuget.org/packages/Auth0.ManagementApi).

Reviewer: ```suggestion
All model classes are now stored in the NuGet package which they relate to. So all Authentication API SDK model classes are stored in the [Auth0.AuthenticationAPI NuGet package](https://www.nuget.org/packages/Auth0.AuthenticationApi). Likewise, all Management API SDK model classes are stored in the [Auth0.ManagementAPI NuGet package](https://www.nuget.org/packages/Auth0.ManagementApi).
```

The [Auth0.Core NuGet package](https://www.nuget.org/packages/Auth0.Core) only contains some shared classes used for communicating with the actual APIs, Exception classes etc.

The separation allows us to evolve these 2 packages in the future more easily in a separate directions.

## Difference between Version 3 and Version 4

> Version 3 of the Auth0.NET SDK can still be used for applications which does not use the new OIDC-conformant pipeline. For these applications you must install the following NuGet packages:

Reviewer: ```suggestion
Version 3 of the Auth0.NET SDK can still be used for applications which do not use the OIDC-conformant pipeline. For these applications you must install the following NuGet packages:
```

* [Auth0.AuthenticationAPI](https://www.nuget.org/packages/Auth0.AuthenticationApi) Version 3.x
* [Auth0.ManagementAPI](https://www.nuget.org/packages/Auth0.ManagementApi) Version 3.x
* [Auth0.Core](https://www.nuget.org/packages/Auth0.Core) Version 3.x

> Version 4 of the Auth0.NET SDK **must be used** for applications which use the new OIDC-conformant pipeline. For these applications you must install the following NuGet packages:

Reviewer: ```suggestion
Version 4 of the Auth0.NET SDK **must be used** for applications which use the OIDC-conformant pipeline. For these applications you must install the following NuGet packages:
```

* [Auth0.AuthenticationAPI](https://www.nuget.org/packages/Auth0.AuthenticationApi) Version 4.x
* [Auth0.ManagementAPI](https://www.nuget.org/packages/Auth0.ManagementApi) Version 4.x
* [Auth0.Core](https://www.nuget.org/packages/Auth0.Core) Version 4.x

## List of changes

Here follows the list of changes made from Version 3 to Version 4 of the Auth0.NET SDK, with guidance on how to change your applications.

### Authentication API

* **Removed** all members previously marked as obsolete. This relates mostly to the methods which did not conform to the *Async naming convention for .NET `async` methods.

> * **Deprecated** the `AuthenticateAsync()` method as the legacy `oauth/ro` endpoint has been disabled. You should use `GetTokenAsync(ResourceOwnerTokenRequest)` instead, and the `AuthenticateAsync()` has been changed to simply call the new `GetTokenAsync(ResourceOwnerTokenRequest)` method. Note that confidential clients will need to provide a `ClientSecret` in addition to the `ClientId`. For more information see the [Resource Owner Password grant type](https://auth0.com/docs/api/authentication#resource-owner-password). 

Reviewer: ```suggestion
* **Deprecated** the `AuthenticateAsync()` method as the legacy `oauth/ro` endpoint has been disabled. You should use `GetTokenAsync(ResourceOwnerTokenRequest)` instead. `AuthenticateAsync()` has been changed to simply call the new `GetTokenAsync(ResourceOwnerTokenRequest)` method. Note that confidential clients will need to provide a `ClientSecret` in addition to the `ClientId`. For more information see the [Resource Owner Password grant type](https://auth0.com/docs/api/authentication#resource-owner-password). 
```

* **Changed** the response of `AuthenticateAsync()` to now return an `AccessTokenResponse` instead of `AuthorizationResponse`. 

> * **Renamed** the `Connection` property in `AuthenticationRequest` class to `Realm`. It is also now optional. If the Connection is not provided in the `Realm` property, the Authentication API will use the connection specified as the Default Directory in the [Account Settings](https://manage.auth0.com/#/account). **As noted above however**, you should rather use `GetTokenAsync(ResourceOwnerTokenRequest)` instead.

Reviewer: ```suggestion
* **Renamed** the `Connection` property in `AuthenticationRequest` class to `Realm`. It is also now optional. If the Connection is not provided in the `Realm` property, the Authentication API will use the connection specified as the Default Directory in the [Account Settings](https://manage.auth0.com/#/account). **As noted above however**, you should use `GetTokenAsync(ResourceOwnerTokenRequest)` instead.
```
